build.py

```diff
@@ -230,10 +230,24 @@ def cuda_offload(mod, args):
     import tvm.relax.backend.contrib.cutlass
     from tvm.relax.backend import get_patterns_with_prefix
 
+    from mlc_llm.transform import combine_parallel_transposed_matmul, rewrite_attention
+
     debug_dump_script(mod, "mod_before_cuda_offload.py", args)
 
+    mod["prefill"] = rewrite_attention(mod["prefill"])
+    mod["decode"] = rewrite_attention(mod["decode"])
```

Comment:
You might be able to remove the custuom attn rewriting and instead use the implicit rewriting done during `partition_for_cutlass`, if you use https://github.com/apache/tvm/pull/14957

Reply:
It doesn't seem to work. My code is at https://github.com/yelite/mlc-llm/blob/attention-test/build.py#L246-L249

Attention isn't rewritten and I can see bindings like
```
            lv1496: R.Tensor((1, m, 32, 128), dtype="float16") = R.reshape(lv1494, R.shape([1, m, 32, 128]))
            lv1497: R.Tensor((1, m, 32, 128), dtype="float16") = R.reshape(lv1495, R.shape([1, m, 32, 128]))
            lv1498: R.Tensor((1, 32, n, 128), dtype="float16") = R.permute_dims(lv1486, axes=[0, 2, 1, 3])
            lv1499: R.Tensor((1, 32, m, 128), dtype="float16") = R.permute_dims(lv1496, axes=[0, 2, 1, 3])
            lv1500: R.Tensor((1, 32, m, 128), dtype="float16") = R.permute_dims(lv1497, axes=[0, 2, 1, 3])
            lv1501: R.Tensor((1, 32, 128, m), dtype="float16") = R.permute_dims(lv1499, axes=[0, 1, 3, 2])
            lv1502: R.Tensor((1, 32, n, m), dtype="float16") = R.matmul(lv1498, lv1501, out_dtype="void")
            lv1503: R.Tensor((1, 32, n, m), dtype="float16") = R.divide(lv1502, metadata["relax.expr.Constant"][125])
            lv1504: R.Tensor((1, 32, n, m), dtype="float16") = R.maximum(lv1503, metadata["relax.expr.Constant"][126])
            lv1505: R.Tensor((1, 32, n, m), dtype="float16") = R.minimum(lv1504, lv5)
            lv1506: R.Tensor((1, 32, n, m), dtype="float32") = R.astype(lv1505, dtype="float32")
            lv1507: R.Tensor((1, 32, n, m), dtype="float32") = R.nn.softmax(lv1506, axis=-1)
            lv1508: R.Tensor((1, 32, n, m), dtype="float16") = R.astype(lv1507, dtype="float16")
            lv1509: R.Tensor((1, 32, n, 128), dtype="float16") = R.matmul(lv1508, lv1500, out_dtype="void")
            lv1510: R.Tensor((1, n, 32, 128), dtype="float16") = R.permute_dims(lv1509, axes=[0, 2, 1, 3])
            lv1511: R.Tensor((1, n, 4096), dtype="float16") = R.reshape(lv1510, R.shape([1, n, 4096]))
            lv1512: R.Tensor((4096, 4096), dtype="float16") = R.permute_dims(linear_weight213, axes=None)
            lv1513: R.Tensor((1, n, 4096), dtype="float16") = R.matmul(lv1511, lv1512, out_dtype="void")
            lv1514: R.Tensor((1, n, 4096), dtype="float16") = R.add(lv1475, lv1513)
```
after rewriting

Reply:
yeah it probably doesn't handle max/min for masking. You can extend the pattern, just like I did in the PR.

Reply:
Currently, we made the following change in the model from the upstream to enable attention offloading: see `mlc_llm/relax_model/llama.py` in https://github.com/mlc-ai/mlc-llm/compare/main...octoml:mlc-llm:main
Would it be possible to extend the pattern to catch the original form?

Reply:
Next week I'll work on improving the implicit rewriter to handle this case. For now this is good to go.

Reply:
But we should use `CombineParallelMatmul` rather than the custom one here. I confirmed that we end with the same number of matmuls. @yelite

Reply:
Sounds good. I have already updated the code to use `CombineParallelMatmul`.